File: backend/app/services/recommendation_engine.py
"""
Recommendation Engine - Generates investment recommendations.
Filters and ranks stocks by strategy: growth, value, defensive, contrarian.
"""
from typing import List, Optional
from app.models.schemas import StockRecommendation, Strategy
from app.services import demo_data
import logging

logger = logging.getLogger(__name__)


class RecommendationEngine:
    """
    Generates and filters stock recommendations based on investment strategies.
    Implements different lenses: growth, value, defensive, contrarian.
    """
    
    def __init__(self):
        self.all_recommendations = demo_data.get_demo_stock_recommendations()
    
    def get_top_opportunities(self, n: int = 10) -> List[StockRecommendation]:
        """
        Returns top N investment opportunities across all sectors.
        Sorted by conviction score.
        """
        sorted_recs = sorted(
            self.all_recommendations,
            key=lambda x: x.conviction_score,
            reverse=True
        )
        return sorted_recs[:n]
    
    def get_sector_opportunities(
        self, 
        sector: str, 
        n: int = 5
    ) -> List[StockRecommendation]:
        """
        Returns best picks within a specific sector.
        """
        sector_recs = [
            rec for rec in self.all_recommendations 
            if rec.sector.lower() == sector.lower()
        ]
        
        sorted_recs = sorted(
            sector_recs,
            key=lambda x: x.conviction_score,
            reverse=True
        )
        
        return sorted_recs[:n]
    
    def get_defensive_picks(self, n: int = 10) -> List[StockRecommendation]:
        """
        Returns defensive/recession-resistant picks.
        
        Criteria:
        - Defensive sectors (Healthcare, Consumer Staples, Utilities)
        - High dividend yield (>2%)
        - Lower volatility (proxied by sector)
        - Strong balance sheets
        
        Philosophy: "Sleep well at night" - Peter Lynch
        """
        defensive_sectors = ['Healthcare', 'Consumer Staples', 'Utilities']
        
        defensive_recs = [
            rec for rec in self.all_recommendations
            if (rec.sector in defensive_sectors or 
                (rec.dividend_yield and rec.dividend_yield > 2.0))
        ]
        
        # Sort by combination of conviction and dividend yield
        sorted_recs = sorted(
            defensive_recs,
            key=lambda x: x.conviction_score + (x.dividend_yield or 0) * 5,
            reverse=True
        )
        
        return sorted_recs[:n]
    
    def get_growth_picks(self, n: int = 10) -> List[StockRecommendation]:
        """
        Returns high-growth momentum plays.
        
        Criteria:
        - Growth sectors (Technology, Healthcare biotech, Communication)
        - High revenue/earnings growth
        - PEG ratio < 2.5 (growth at reasonable price - Peter Lynch GARP)
        - Strong secular tailwinds
        
        Philosophy: "The best business to own is one that over time can employ 
        large amounts of capital at very high rates of return" - Warren Buffett
        """
        growth_sectors = [
            'Technology', 
            'Communication Services', 
            'Healthcare',
            'Consumer Discretionary'
        ]
        
        growth_recs = [
            rec for rec in self.all_recommendations
            if (rec.sector in growth_sectors and
                rec.peg_ratio and rec.peg_ratio < 3.0 and
                len(rec.tailwinds) >= 3)
        ]
        
        # Sort by conviction score
        sorted_recs = sorted(
            growth_recs,
            key=lambda x: x.conviction_score,
            reverse=True
        )
        
        return sorted_recs[:n]
    
    def get_value_picks(self, n: int = 10) -> List[StockRecommendation]:
        """
        Returns deep value/turnaround candidates.
        
        Criteria:
        - Low P/E ratio (<15)
        - Significant upside to fair value (>15%)
        - Out of favor but fundamentals improving
        - Margin of safety
        
        Philosophy: "Price is what you pay, value is what you get" - Warren Buffett
        """
        value_recs = [
            rec for rec in self.all_recommendations
            if (rec.pe_ratio and rec.pe_ratio < 20 and
                rec.upside_potential_pct > 10)
        ]
        
        # Sort by upside potential (margin of safety)
        sorted_recs = sorted(
            value_recs,
            key=lambda x: x.upside_potential_pct,
            reverse=True
        )
        
        return sorted_recs[:n]
    
    def get_contrarian_picks(self, n: int = 10) -> List[StockRecommendation]:
        """
        Returns contrarian/oversold opportunities with improving fundamentals.
        
        Criteria:
        - Sectors currently out of favor (bearish outlook)
        - More tailwinds than headwinds emerging
        - Significant upside potential
        - Turning points in fundamentals
        
        Philosophy: "The time to buy is when there's blood in the streets" - Baron Rothschild
        "You make money by buying things that are down and out" - Howard Marks
        """
        # Look for stocks with strong upside despite bearish sector sentiment
        contrarian_recs = [
            rec for rec in self.all_recommendations
            if (rec.upside_potential_pct > 15 and
                len(rec.tailwinds) > len(rec.headwinds))
        ]
        
        # Sort by upside potential
        sorted_recs = sorted(
            contrarian_recs,
            key=lambda x: x.upside_potential_pct,
            reverse=True
        )
        
        return sorted_recs[:n]
    
    def filter_recommendations(
        self,
        strategy: Optional[Strategy] = Strategy.ALL,
        sector: Optional[str] = None,
        min_conviction: Optional[float] = None,
        limit: int = 20
    ) -> List[StockRecommendation]:
        """
        Filters recommendations based on multiple criteria.
        
        Args:
            strategy: Investment strategy filter (growth/value/defensive/contrarian)
            sector: Sector filter
            min_conviction: Minimum conviction score
            limit: Maximum number of results
        """
        # Start with appropriate strategy subset
        if strategy == Strategy.GROWTH:
            filtered = self.get_growth_picks(n=limit)
        elif strategy == Strategy.VALUE:
            filtered = self.get_value_picks(n=limit)
        elif strategy == Strategy.DEFENSIVE:
            filtered = self.get_defensive_picks(n=limit)
        elif strategy == Strategy.CONTRARIAN:
            filtered = self.get_contrarian_picks(n=limit)
        else:
            filtered = self.all_recommendations.copy()
        
        # Apply sector filter
        if sector:
            filtered = [
                rec for rec in filtered
                if rec.sector.lower() == sector.lower()
            ]
        
        # Apply conviction filter
        if min_conviction:
            filtered = [
                rec for rec in filtered
                if rec.conviction_score >= min_conviction
            ]
        
        # Sort by conviction and limit
        sorted_recs = sorted(
            filtered,
            key=lambda x: x.conviction_score,
            reverse=True
        )
        
        return sorted_recs[:limit]
    
    def get_recommendation_by_ticker(self, ticker: str) -> Optional[StockRecommendation]:
        """
        Returns detailed recommendation for a specific ticker.
        """
        for rec in self.all_recommendations:
            if rec.ticker.upper() == ticker.upper():
                return rec
        return None
    
    def get_portfolio_ideas(
        self,
        risk_tolerance: str = "moderate",
        n: int = 10
    ) -> List[StockRecommendation]:
        """
        Returns a diversified portfolio of recommendations based on risk tolerance.
        
        Conservative: Heavy defensive, income focus
        Moderate: Balanced across strategies
        Aggressive: Growth and cyclical focus
        """
        if risk_tolerance.lower() == "conservative":
            # 60% defensive, 40% quality growth
            defensive = self.get_defensive_picks(n=int(n * 0.6))
            quality_growth = self.get_growth_picks(n=int(n * 0.4))
            return defensive + quality_growth
        
        elif risk_tolerance.lower() == "aggressive":
            # 60% growth, 30% contrarian, 10% value
            growth = self.get_growth_picks(n=int(n * 0.6))
            contrarian = self.get_contrarian_picks(n=int(n * 0.3))
            value = self.get_value_picks(n=int(n * 0.1))
            return growth + contrarian + value
        
        else:  # moderate
            # 40% growth, 30% defensive, 20% value, 10% contrarian
            growth = self.get_growth_picks(n=int(n * 0.4))
            defensive = self.get_defensive_picks(n=int(n * 0.3))
            value = self.get_value_picks(n=int(n * 0.2))
            contrarian = self.get_contrarian_picks(n=int(n * 0.1))
            return growth + defensive + value + contrarian

File: backend-ts/src/services/recommendationEngine.ts
// Recommendation Engine - Generate investment recommendations

import { supabase, getRecommendations, getSectors } from '../config/supabase';
import { logger } from '../utils/logger';
import { Recommendation, Sector, RecommendationFilters } from '../types';

export class RecommendationEngine {
  /**
   * Get top investment opportunities across all sectors
   */
  async getTopOpportunities(limit: number = 10): Promise<Recommendation[]> {
    try {
      const recommendations = await getRecommendations();
      
      // Sort by conviction score
      const sorted = recommendations.sort((a, b) => 
        b.conviction_score - a.conviction_score
      );

      return sorted.slice(0, limit);
    } catch (error: any) {
      logger.error('Error getting top opportunities', error.message);
      throw error;
    }
  }

  /**
   * Get best picks within a specific sector
   */
  async getSectorOpportunities(sectorName: string, limit: number = 5): Promise<Recommendation[]> {
    try {
      const recommendations = await getRecommendations({ sector: sectorName });
      
      // Sort by conviction score
      const sorted = recommendations.sort((a, b) => 
        b.conviction_score - a.conviction_score
      );

      return sorted.slice(0, limit);
    } catch (error: any) {
      logger.error(`Error getting sector opportunities for ${sectorName}`, error.message);
      throw error;
    }
  }

  /**
   * Get defensive/recession-resistant picks
   * 
   * Criteria:
   * - Defensive sectors (Healthcare, Consumer Staples, Utilities)
   * - Lower risk level
   * - Strong conviction scores
   * 
   * Philosophy: "Sleep well at night" - Peter Lynch
   */
  async getDefensivePicks(limit: number = 10): Promise<Recommendation[]> {
    try {
      const recommendations = await getRecommendations();
      
      const defensiveSectors = ['Healthcare', 'Consumer Staples', 'Utilities'];
      
      const defensiveRecs = recommendations.filter(rec => {
        const sectorName = rec.sectors?.name || '';
        return defensiveSectors.includes(sectorName) || rec.risk_level === 'low';
      });

      // Sort by conviction score
      const sorted = defensiveRecs.sort((a, b) => 
        b.conviction_score - a.conviction_score
      );

      return sorted.slice(0, limit);
    } catch (error: any) {
      logger.error('Error getting defensive picks', error.message);
      throw error;
    }
  }

  /**
   * Get high-growth momentum plays
   * 
   * Criteria:
   * - Growth sectors (Technology, Healthcare, Communication)
   * - Growth strategy
   * - Strong catalysts (3+)
   * 
   * Philosophy: "The best business to own is one that over time can employ 
   * large amounts of capital at very high rates of return" - Warren Buffett
   */
  async getGrowthPicks(limit: number = 10): Promise<Recommendation[]> {
    try {
      const recommendations = await getRecommendations({ strategy: 'growth' });
      
      const growthSectors = [
        'Technology', 
        'Communication Services', 
        'Healthcare',
        'Consumer Discretionary'
      ];

      const growthRecs = recommendations.filter(rec => {
        const sectorName = rec.sectors?.name || '';
        const hasCatalysts = rec.catalysts && rec.catalysts.length >= 3;
        return growthSectors.includes(sectorName) && hasCatalysts;
      });

      // Sort by conviction score
      const sorted = growthRecs.sort((a, b) => 
        b.conviction_score - a.conviction_score
      );

      return sorted.slice(0, limit);
    } catch (error: any) {
      logger.error('Error getting growth picks', error.message);
      throw error;
    }
  }

  /**
   * Get deep value/turnaround candidates
   * 
   * Criteria:
   * - Value strategy
   * - Significant upside to fair value (>10%)
   * - Out of favor but fundamentals improving
   * 
   * Philosophy: "Price is what you pay, value is what you get" - Warren Buffett
   */
  async getValuePicks(limit: number = 10): Promise<Recommendation[]> {
    try {
      const recommendations = await getRecommendations({ strategy: 'value' });
      
      const valueRecs = recommendations.filter(rec => {
        const pe = rec.valuation_metrics?.pe || 0;
        return pe < 20 && rec.upside_percent > 10;
      });

      // Sort by upside potential (margin of safety)
      const sorted = valueRecs.sort((a, b) => 
        b.upside_percent - a.upside_percent
      );

      return sorted.slice(0, limit);
    } catch (error: any) {
      logger.error('Error getting value picks', error.message);
      throw error;
    }
  }

  /**
   * Get contrarian/oversold opportunities with improving fundamentals
   * 
   * Criteria:
   * - Significant upside potential
   * - More catalysts than risks
   * - Contrarian strategy
   * 
   * Philosophy: "The time to buy is when there's blood in the streets" - Baron Rothschild
   */
  async getContrarianPicks(limit: number = 10): Promise<Recommendation[]> {
    try {
      const recommendations = await getRecommendations({ strategy: 'contrarian' });
      
      const contrarianRecs = recommendations.filter(rec => {
        const catalystCount = rec.catalysts?.length || 0;
        const riskCount = rec.risks?.length || 0;
        return rec.upside_percent > 15 && catalystCount > riskCount;
      });

      // Sort by upside potential
      const sorted = contrarianRecs.sort((a, b) => 
        b.upside_percent - a.upside_percent
      );

      return sorted.slice(0, limit);
    } catch (error: any) {
      logger.error('Error getting contrarian picks', error.message);
      throw error;
    }
  }

  /**
   * Filter recommendations based on multiple criteria
   */
  async filterRecommendations(
    filters: RecommendationFilters,
    limit: number = 20
  ): Promise<Recommendation[]> {
    try {
      // Get strategy-specific picks
      let recommendations: Recommendation[];

      if (filters.strategy === 'growth') {
        recommendations = await this.getGrowthPicks(limit * 2);
      } else if (filters.strategy === 'value') {
        recommendations = await this.getValuePicks(limit * 2);
      } else if (filters.strategy === 'defensive') {
        recommendations = await this.getDefensivePicks(limit * 2);
      } else if (filters.strategy === 'contrarian') {
        recommendations = await this.getContrarianPicks(limit * 2);
      } else {
        // All strategies
        recommendations = await getRecommendations(filters);
      }

      // Apply additional filters
      let filtered = recommendations;

      if (filters.minConviction) {
        filtered = filtered.filter(rec => rec.conviction_score >= filters.minConviction!);
      }

      if (filters.sector) {
        filtered = filtered.filter(rec => {
          const sectorName = rec.sectors?.name || '';
          return sectorName.toLowerCase() === filters.sector!.toLowerCase();
        });
      }

      if (filters.riskLevel) {
        filtered = filtered.filter(rec => rec.risk_level === filters.riskLevel);
      }

      return filtered.slice(0, limit);
    } catch (error: any) {
      logger.error('Error filtering recommendations', error.message);
      throw error;
    }
  }

  /**
   * Analyze sector performance and generate insights
   */
  async analyzeSector(sectorName: string): Promise<any> {
    try {
      const { data: sector, error: sectorError } = await supabase
        .from('sectors')
        .select('*')
        .eq('name', sectorName)
        .single();

      if (sectorError) throw sectorError;

      const recommendations = await this.getSectorOpportunities(sectorName, 10);

      // Calculate sector statistics
      const avgConviction = recommendations.length > 0
        ? recommendations.reduce((sum, rec) => sum + rec.conviction_score, 0) / recommendations.length
        : 0;

      const avgUpside = recommendations.length > 0
        ? recommendations.reduce((sum, rec) => sum + rec.upside_percent, 0) / recommendations.length
        : 0;

      return {
        sector,
        recommendations,
        statistics: {
          totalRecommendations: recommendations.length,
          averageConviction: avgConviction.toFixed(2),
          averageUpside: avgUpside.toFixed(2),
          highConvictionCount: recommendations.filter(r => r.conviction_score >= 8.0).length,
          lowRiskCount: recommendations.filter(r => r.risk_level === 'low').length
        }
      };
    } catch (error: any) {
      logger.error(`Error analyzing sector ${sectorName}`, error.message);
      throw error;
    }
  }
}
